### PR TITLE
[PropertyInfo] Add accessor and mutator extractor interface and implementation on reflection

### PR DESCRIPTION
**Pull request by joelwurtz**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no    
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #30248, partially: #22190, #18016, #5013, #9336, #5219,
| License       | MIT
| Doc PR        | TODO

This PR brings accessor / mutator extraction on the PropertyInfo component,

There is no link to existing code, as IMO it should be in another PR as this will add a dependency on property access to the property info component and not sure this is something wanted (although, it will reduce a lot of code base on the property access component as a lot of code seems to be duplicated)

Code is extracted from #30248 also there is some new features (that can be removed if not wanted)

 * Allow extracting private accessor / mutator (will do a new PR that improve private extraction on reflection latter)
 * Allow extracting static accessor / mutators
 * Allow extracting constructor mutators

Current implementation try to be as close as the PropertyAccess implementation and i did not reuse some methods already available in the class as there is some differences in implementation, but maybe it will be a good time to make this consistent (Looking forward to your input) ?

Things that should be done in a new PR:

 * Linking property info to property access to remove a lot of duplicate code
 * Add a new system that allow adding Virtual Property based on this extractor

### CODE REVIEW

**Comment by joelwurtz** [comment]
#30706 add the private property concept, if this gets merged will update this PR to handle the flags options when extracting mutator / accessor

**Comment by fabpot** [comment]
@joelwurtz #30706 has just been merged :)

**Comment by joelwurtz** [comment]
Thanks @fabpot just updated the PR to reflect the change and extracting accessor / mutator with access flags.

I also begin to add consistency to the extraction, but not sure this is wanted, here is a list of different behavior between `ReflectionExtractor` and `PropertyAccessor` class where have i a question about what do to:

 * `ReflectionExtractor` use a configurable list of accessor and mutator prefixes to know if the property exist and can be read / write where `PropertyAccessor` has a hard coded list: I don't think it's a problem as default list is the same as the hard coded list of `PropertyAccessor`: Should we use the `ReflectionExtractor` list and not the hard coded one ?

 * `PropertyAccessor` allows to get / set on methods having the same name (jQuery style `public function foo(string $foo = null): string)`), where `ReflectionExtractor` does not allow it, and adding an empty prefix to the list breaks some tests, Should we add support for this or change the code to allow empty prefix ?

 * `PropertyAccessor` limit add / remove mutator only when type of the property is a traversable / array and only when **both** methods are available, where `ReflectionExtractor` does the reverse, it suppose that type is array if the method is present, and made the property writable if **one** of the `add` or `remove` method if available: Should we add support for restricting those mutators like `PropertyAccessor` do ? 

 * `PropertyAccessor` has support for magic call where `ReflectionExtractor` does not have it by default (private property + __get or __set = not writable or readable): Should we add support for magic set / get in `ReflectionExtractor` ?

Support for each of this behavior can be done in constructor arguments or context, would prefer to avoid context and, in the future, use different `ReflectionExtractor` service for each use case, as most of the time those behavior are domain specific (for a set of class, not for one only).

**Comment by dunglas** [comment]
I fully agree with this move. I also think we should make `PropertyInfo` a hard dependency of `PropertyAccess` (and `PropertyAccess` a hard dependency of `Serializer`) prevent code duplication. It's ok for Symfony components to depend of other lower level Symfony components.

> Should we use the ReflectionExtractor list and not the hard coded one ?

I think so.

> Should we add support for this or change the code to allow empty prefix.

I'm not sure that the jQuery-style is still very used. jQuery itself is not used anymore in modern JS projects. It's why I not implemented it (and AFAIK nobody complained since the creation of this library). As a an alternative, we couldn't just deprecate support for this notation. WDYT?

> Should we add support for restricting those mutators like PropertyAccessor do ?

It looks reasonable.

> Should we add support for magic set / get?

Yes, it could be nice as long as it's optional. I'm not sure that this feature is used a lot, but it could be useful for advanced use cases.

**Comment by fabpot** [comment]
I agree with @dunglas, that having hard dependencies between such components make sense (especially as they are low-level building blocks).

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    public function isStatic(): bool
+    {
+        return $this->static;
+    }
+
+    public function getConstructorParameter(): ?\ReflectionParameter
+    {
+        return $this->constructorParameter;
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: this seems a bit fragile to me. should we check the type and throw an exception if its not TYPE_CONSTRUCTOR? then we could make the return type of this non-nullable. (and also validate in the constructor)

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * Writes mutator tell how to write to a property.
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+final class WriteMutator
+{
+    public const TYPE_METHOD = 1;
+    public const TYPE_PROPERTY = 2;
+    public const TYPE_ARRAY_DIMENSION = 3;
+    public const TYPE_ADDER_REMOVER_METHOD = 4;
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: are the adder and remover exclusively for this case?

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+namespace Symfony\Component\PropertyInfo;
+
+/**
+ * Writes mutator tell how to write to a property.
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+final class WriteMutator
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: what about collection fields?

do we have type information about the property, or is that out of scope for this?

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    private $type;
+    private $name;
+    private $adderName;
+    private $removerName;
+    private $private;
+    private $constructorParameter;
+    private $static;
+
+    public function __construct(int $type, ?string $name, ?string $adderName, ?string $removerName, bool $private = false, bool $static = false, \ReflectionParameter $constructorParameter = null)
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: this is a bit fragile. we could make the constructor private and have static create<Type> methods with all required arguments for that type.

```php
public static createMethod(string $methodName); // could this be private?
public static createProperty(string $propertyName, bool $private, bool $private); 
public static createArrayDimension(string $key);
public static createAdderRemoverMethod(string $adderName, string $removerName);
public static createConstructorArgument($constructorParameter);

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    public function isPrivate(): bool
+    {
+        return $this->private;
+    }
+
+    public function isStatic(): bool
+    {
+        return $this->static;
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: is this applicable for methods as well, or only for the property? i think we should throw an exception when we are a type that does not have it.

**Review by dbu — COMMENTED** [review]
nice, i like these very limited information models as building blocks.

i guess a serializer would use these as one element along with other things like the target name (serializer) resp source name (deserialize)?

i commented some random thoughts i had while reading this. lets discuss tomorrow.

File: src/Symfony/Component/PropertyInfo/ReadAccessor.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        $this->private = $private;
+        $this->static = $static;
+    }
+
+    public function getType(): int
+    {
+        return $this->type;
+    }
+
+    public function getName(): string
```
[the remark this answers is not in the document]
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/ReadAccessor.php]: Yes it needs doc,

> for a serializer, you will also care about the name in the serialized form (so even if you use a getter, you care about the actual name of the property). but that is metadata for the serializer that is out of scope here. but to illustrate that there is more than one "name" thing.

It's the name relative to how to access the property, the name exposed is the one to use, other names are implementation relative and should not be exposed.

File: src/Symfony/Component/PropertyInfo/ReadAccessorExtractorInterface.php [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyInfo;
+
+/**
+ * Extract read accessor for property of a class.
```
[the remark this answers is not in the document]
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/ReadAccessorExtractorInterface.php]: > Extract read accessor information for the property of a class.

:+1: 

about `Extractor` word this is mainly for consistency with other class / interface

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    public function getConstructorParameter(): ?\ReflectionParameter
```
[the remark this answers is not in the document]
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: This is what i'm using in the AutoMapper component and when it rewrite this i was wondering the same question, i am also in favor of not exposing the reflection parameter here as it's a detail of implementation, but need to rebase the AutoMapper component to see what info i need here.

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 5 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: applicable for boths, so no need for exception

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 4 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: Not sure if applicable: Need to test this with how the mutator / accessor is created in ReflectionExtractor.

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 2 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: You mean the method ?

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 1 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: Why not, would prefer to not have a lot of logic here, for me this class is mainly a helper for IDE and stuff that avoid using an array like in the PropertyAccess component.

Adding exception can have some bias in other component where we would need to take care of this (more maintenance)

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 3 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: Type information is out of scope, it should be extracted using a PropertyTypeExtractor

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 5 above)
Comment (dbu) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: it makes no sense for the constructor type.

i'd love to make these methods strict to find programming errors early when using this.

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 2 above)
Comment (dbu) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: sorry, yes, i mean whether those two method names are only for this case, or also for property / method

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (dbu) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: does a reflection parameter determine the position of the constructor argument? if it does or has some other relevant upside over the string name, i think its ok to have this here.
 
 if not, i would prefer a string, and if the AutoMapper needs to extract the reflection for BC it can do that iself, rather than have a new component do unnecessary things to make something else easier to allow BC.

**Comment by joelwurtz** [comment]
Thanks @dbu for the review

> i guess a serializer would use these as one element along with other things like the target name (serializer) resp source name (deserialize)?

With the global vision, serializer will never use this :)

It intents to work as the following schema:

 * Serializer use the PropertyAccess component to read / write a value for a data
 * PropertyAccess use this extractor to know how to read / write a value for an object

This will mask the complexity of accessing / setting value on the serializer component and also ensure that we can also read / write to / from an array or other dynamic object that does not have metadatas.

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 1 above)
Comment (dbu) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: if you want to be sure that your code won't fatal in random places, having optional return value is a huge pain as you need to check if you actually got something. (thats why phpstan will complain)

imho an exception when you do something you are not supposed to do would be more helpful.

i agree we should not have too much logic on these, but we have models that cover different cases and we should make sure that the cases do not get mixed up. the super-duper OO alternative would be some marker interface and an implementation for each type, but that seems overcomplicated to me, and makes it harder to see which types exist.

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 3 above)
Comment (dbu) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: but does the WriteMutator help with how to write to a collection field, or would that be up to the consumer to figure out? either there is add/remove or there is a property that is a collection that implements those methods... what if i should use a getter to get the collection and then modify it?

**Comment by dbu** [comment]
> With the global vision, serializer will never use this

thanks for the clarifications. and i think that makes sense. the only concern i have is with performance. large object graphs with many fields can mean A LOT of method calls. if there are several 10k of single fields to handle, method calls for each property start to add up. this is what prompted us to write https://github.com/liip/serializer/ . 
i agree that for the normal case, a bit of overhead is worth it when it avoids a lot of duplication in the codebase. but i'd love to check how we can make code generation possible so that it can cleanly replace the slower implementation keeping the exact same behaviour.

**Comment by joelwurtz** [comment]
> but i'd love to check how we can make code generation possible so that it can cleanly replace the slower implementation keeping the exact same behaviour.

This is clearly something that i want to make it possible with new organisation / extension point. As we could do a specific PropertyAccess implementation that generate code given accessor / mutator (see the automapper pull request https://github.com/symfony/symfony/pull/30248 that have something like this), which will boost serializer by a large factor.

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 2 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: only for this case

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 3 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: This is up to the consumer to figure out the best way to do that

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: This has been remove on this PR, will readd it if needed, but i don't think it's useful in the Automapper

File: src/Symfony/Component/PropertyInfo/WriteMutator.php (hunk 4 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: WriteMutator / ReadAccessor now use static constructor for better UX

**Comment by joelwurtz** [comment]
Ready for review

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 9]
```diff
@@ -209,6 +194,140 @@ public function isInitializable(string $class, string $property, array $context
...
+     */
+    public function getReadAccessor(string $class, string $property, array $context = []): ?ReadAccessor
+    {
+        try {
+            $reflClass = new \ReflectionClass($class);
+        } catch (\ReflectionException $e) {
+            return null;
+        }
+
+        $allowGetterSetter = $context['enable_getter_setter_extraction'] ?? false;
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: i'd prefer constants over magic strings for options

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        return $this->type;
+    }
+
+    public function getName(): ?string
+    {
+        return $this->name;
+    }
+
+    public function getAdder(): ?self
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: this (and remover) would be of TYPE_METHOD only, right? we should state that explicitly.

and i still think we should throw exceptions when the WriteMutator is not TYPE_ADDER_REMOVER_METHOD

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    {
+        return $this->adder;
+    }
+
+    public function getRemover(): ?self
+    {
+        return $this->remover;
+    }
+
+    public function isPrivate(): ?bool
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: i'd prefer to throw an exception when the mutator is of a type where this information makes no sense (like constructor)

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    private function __construct()
+    {
+    }
+
+    public function getType(): int
+    {
+        return $this->type;
+    }
+
+    public function getName(): ?string
```
Comment (dbu) [inline src/Symfony/Component/PropertyInfo/WriteMutator.php]: i'd throw an exception when this is TYPE_ADDER_REMOVER_METHOD and then we can say we always return string.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 13]
```diff
@@ -390,17 +394,22 @@ private function readProperty($zval, $property, bool $ignoreInvalidProperty = fa
...
+                $result[self::VALUE] = $object->{$access->getName()};
 
-            if ($access[self::ACCESS_REF] && isset($zval[self::REF])) {
-                $result[self::REF] = &$object->{$access[self::ACCESS_NAME]};
+                if (isset($zval[self::REF]) && $access->canBeReference()) {
+                    $result[self::REF] = &$object->{$access->getName()};
+                }
             }
-        } elseif (!$access[self::ACCESS_HAS_PROPERTY] && property_exists($object, $property)) {
+        } elseif ($object instanceof \stdClass && property_exists($object, $property)) {
```
Comment (dbu) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: i we could refactor this to do early return instead of all the elseif, while we are touching this

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 14]
```diff
@@ -390,17 +394,22 @@ private function readProperty($zval, $property, bool $ignoreInvalidProperty = fa
 
         $result = self::$resultProto;
         $object = $zval[self::VALUE];
-        $access = $this->getReadAccessInfo(\get_class($object), $property);
+        $class = \get_class($object);
+        $access = $this->getReadAccessor($class, $property);
+
+        if (null !== $access) {
+            if (ReadAccessor::TYPE_METHOD === $access->getType()) {
+                $result[self::VALUE] = $object->{$access->getName()}();
```
Comment (dbu) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: you should early return here. the sequence of `if` that are mutually exclusive is confusing. otherwise we could do a switch, that would also make sense.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 13 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Early return may change behavior ? If so, It's a BC break, and would prefer to not touch this

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 14 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: We cannot do early return as we handle zval if the passed value is a reference

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 13 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: We talk about this, and we prefer to not change this here, as it may have some undefined behavior. And would be better to do it in another PR.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php (hunk 14 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: We talk about this, and we prefer to not change this here, as it may have some undefined behavior. And would be better to do it in another PR.

**Comment by joelwurtz** [comment]
ping @Simperfit

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php (hunk 9 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: I also want to use constant in the long term, not sure where to put them however :s

**Comment by dbu** [comment]
for the most difficult question about naming things: i don't like the name "extractor" because that is mainly suiting for reflection. however, in a serializer, the mutator and reader information could be determined from annotations or external metadata like yaml or xml.

File: src/Symfony/Component/PropertyInfo/AccessorExtractorInterface.php [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Symfony\Component\PropertyInfo;
+
+/**
+ * Extracts read accessor and write mutator for property of a class.
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+interface AccessorExtractorInterface extends ReadAccessorExtractorInterface, WriteMutatorExtractorInterface
```
Comment (webmozart) [inline src/Symfony/Component/PropertyInfo/AccessorExtractorInterface.php]: Does this interface make sense? PropertyInfo has individual interfaces for specific purposes and one common `PropertyInfoExtractorInterface` implementing all of the individual ones. Unless you have a specific reason for introducing this intermediate interface, I would remove it.

**Review by webmozart — CHANGES_REQUESTED** [review]
Thanks for your work on this! I have the following concerns:

* If the goal is to provide *all* information that is required in order to read/write a field, there's quite a bit of information missing. 
    * For constructors, we don't know at what position the property should be inserted and if it is nullable
    * For setters, we don't know at what position the property should be inserted. If we rehaul all of this in order to support DDD, we should support writing of multiple properties through the same setter.
    * Both setters and getters could have parameters with default values at intermediate positions. E.g. a setter could have two parameters, both with a default value, and I only want to write the second one.
    * When talking about setters and getters, we should probably anticipate native support for getters and setters in PHP and choose our naming appropriately. Native setters and getters are not around the corner, but this component is here to stay and it will be a pain to change this later on.
* The naming is still a bit inconsistent. I left suggestions for different consistent strategies below.

Overall, I'm not sure what the goal is here.

* If the goal is to provide a more powerful backbone for the Serializer and PropertyAccessor, then this does not improve on the current situation (i.e. being less limited than the Serializer already is, i.e. methods with multiple arguments etc.) and should be thought through more to do so.

* If the goal is to restructure PropertyAccess and maybe remove some duplicated code, then this probably works, but then is it really worth slowing PropertyAccess down just for architectural purity?

**Comment by webmozart** [comment]
Another thought: One limitation of PropertyAccess, which I don't particularly like, is that it can only write properties one by one. As soon as you want to change multiple properties through one method, or pass multiple properties to the constructor, you can't.

Is this really an architecture we want to build to Serializer on?

File: src/Symfony/Component/PropertyInfo/WriteMutator.php [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        $mutator = new self();
+        $mutator->type = self::TYPE_ADDER_REMOVER_METHOD;
+        $mutator->adder = $adder;
+        $mutator->remover = $remover;
+
+        return $mutator;
+    }
+
+    public static function createConstructor(string $propertyName)
```
[the remark this answers is not in the document]
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/WriteMutator.php]: ATM it only need this information in the #30248 PR, will add new information as needed or can change signature since it's now internal.

**Comment by joelwurtz** [comment]
Thanks @webmozart for the review

I add `@internal` on those interface / value object, as we don't provide actually extension point for this and like we discuss it would be better ATM to use that only as an implementation detail to ease code between components.

We could remove `@internal` in another PR (if someone wants to use virtual setter / getter in its project by example), and at this moment we could have a longer discussion about exposed API and value object. Let's keep it simple for this PR.

> Is this really an architecture we want to build the Serializer on?

Like we discuss, this is already the architecture of Normalizer, so we need to keep that in order to not break too many behaviours. 

Another architecture process is done by the #30248 PR which is better IMO, and maybe we would be able to provide a migrate path for the future, but this is clearly not in the scope of this PR.

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 17]
```diff
@@ -209,6 +195,135 @@ public function isInitializable(string $class, string $property, array $context
...
+
+        if ($allowAdderRemover && null !== $methods = $this->findAdderAndRemover($reflClass, (array) Inflector::singularize($camelized))) {
+            [$adderAccessName, $removerAccessName] = $methods;
+
+            $adderMethod = $reflClass->getMethod($adderAccessName);
+            $removerMethod = $reflClass->getMethod($removerAccessName);
+
+            return PropertyWriteInfo::forAdderAndRemover(
+                PropertyWriteInfo::forSetter($adderAccessName, $this->getWriteVisiblityForMethod($adderMethod), $adderMethod->isStatic()),
+                PropertyWriteInfo::forSetter($removerAccessName, $this->getWriteVisiblityForMethod($removerMethod), $removerMethod->isStatic())
```
Comment (webmozart) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: I understand now why this was called "method" and not "setter" before.

* A write info of type "method" is an arbitrary method
* A write info of type "adder and remover" is a composition of two "method" write infos
* A write info of type "constructor" is simplified and does not bear a method name because the method name is implicit

Given that, I think "method" was indeed a better choice of naming than "setter" and "getter". So I think reverting this to `forMethod()` and `TYPE_METHOD` (same for read info) makes sense.

Sorry for the back and forward. :)

File: src/Symfony/Component/PropertyInfo/PropertyWriteInfo.php [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+    {
+        $mutator = new self();
+        $mutator->type = self::TYPE_ADDER_AND_REMOVER;
+        $mutator->adderInfo = $adder;
+        $mutator->removerInfo = $remover;
+
+        return $mutator;
+    }
+
+    public static function forConstructor(string $propertyName): self
```
Comment (webmozart) [inline src/Symfony/Component/PropertyInfo/PropertyWriteInfo.php]: Should we have a visibility here as well? Constructors could be private.

**Review by webmozart — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/PropertyInfo/PropertyWriteInfo.php (hunk 18 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/PropertyWriteInfo.php]: I'm hesitating for that, constructor can be private / protected or public, but having the visiblity info on the property makes little sense (since the visiblity is on the method, not the property).

Can we keep this like that for the moment ? I admit the constructor thing is not the best ATM but would like to change signature and information when it will be used (it's only for the #30248 PR ATM) ?

Comment (webmozart) [reply src/Symfony/Component/PropertyInfo/PropertyWriteInfo.php]: :+1:

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php (hunk 17 above)
Comment (joelwurtz) [reply src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: Fixed

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 19]
```diff
@@ -845,3 +687,7 @@ public static function createCache($namespace, $defaultLifetime, $version, Logge
         return $apcu;
     }
 }
+
+class Popo extends \stdClass
```
Comment (maxhelias) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: I think it should not be here, no ?

**Comment by Korbeil** [comment]
It's finally ready, ping @dunglas 
Last failed test is same as `master` and not related.

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 20]
```diff
@@ -78,6 +82,13 @@ public function __construct(bool $magicCall = false, bool $throwExceptionOnInval
         $this->ignoreInvalidIndices = !$throwExceptionOnInvalidIndex;
         $this->cacheItemPool = $cacheItemPool instanceof NullAdapter ? null : $cacheItemPool; // Replace the NullAdapter by the null value
         $this->ignoreInvalidProperty = !$throwExceptionOnInvalidPropertyPath;
+        $this->readInfoExtractor = $this->writeInfoExtractor = new ReflectionExtractor(
```
Comment (lyrixx) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: Why do you use 2 differents properties?

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 21]
```diff
@@ -78,6 +82,13 @@ public function __construct(bool $magicCall = false, bool $throwExceptionOnInval
...
+            ['set'],
+            ['get', 'is', 'has', 'can'],
+            ['add', 'remove'],
```
Comment (lyrixx) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: You can use `ReflectionExtractor::$defaultArrayMutatorPrefixes`

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 22]
```diff
@@ -78,6 +82,13 @@ public function __construct(bool $magicCall = false, bool $throwExceptionOnInval
         $this->ignoreInvalidIndices = !$throwExceptionOnInvalidIndex;
         $this->cacheItemPool = $cacheItemPool instanceof NullAdapter ? null : $cacheItemPool; // Replace the NullAdapter by the null value
         $this->ignoreInvalidProperty = !$throwExceptionOnInvalidPropertyPath;
+        $this->readInfoExtractor = $this->writeInfoExtractor = new ReflectionExtractor(
+            ['set'],
+            ['get', 'is', 'has', 'can'],
```
Comment (lyrixx) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: If the order does not matter, you can use `ReflectionExtractor::$defaultAccessorPrefixes`

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 23]
```diff
@@ -78,6 +82,13 @@ public function __construct(bool $magicCall = false, bool $throwExceptionOnInval
...
+            ['add', 'remove'],
+            false,
+            ReflectionExtractor::ALLOW_PUBLIC
```
Comment (lyrixx) [inline src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: This is the default value, it can be removed

File: src/Symfony/Component/PropertyInfo/CHANGELOG.md [hunk 24]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.1.0
+-----
+
+* Add accessor and mutator extractor interface and implementation on reflection
```
Comment (lyrixx) [inline src/Symfony/Component/PropertyInfo/CHANGELOG.md]: Add support for extracting accessor and mutator via PHP Reflection

File: src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorCollectionTest.php [hunk 25]
```diff
@@ -188,7 +188,7 @@ public function testIsWritableReturnsFalseIfNoAdderNorRemoverExists()
     public function testSetValueFailsIfAdderAndRemoverExistButValueIsNotTraversable()
     {
         $this->expectException('Symfony\Component\PropertyAccess\Exception\NoSuchPropertyException');
-        $this->expectExceptionMessageRegExp('/Could not determine access type for property "axes" in class "Symfony\\\\Component\\\\PropertyAccess\\\\Tests\\\\PropertyAccessorCollectionTest_Car[^"]*": The property "axes" in class "Symfony\\\\Component\\\\PropertyAccess\\\\Tests\\\\PropertyAccessorCollectionTest_Car[^"]*" can be defined with the methods "addAxis\(\)", "removeAxis\(\)" but the new value must be an array or an instance of \\\\Traversable, "string" given./');
+        $this->expectExceptionMessageRegExp('/Could not determine access type for property "axes" in class "Symfony\\\Component\\\PropertyAccess\\\Tests\\\PropertyAccessorCollectionTest_Car"./');
```
Comment (dunglas) [inline src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorCollectionTest.php]: The previous error messages were more explicit and more helpful for debugging. We should preserve all of them.

File: src/Symfony/Component/PropertyInfo/PropertyReadInfo.php [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * Whether this accessor can be accessed by reference.
+     */
+    public function canBeReference(): bool
+    {
+        return $this->byRef;
+    }
+
+    public static function forProperty(string $propertyName, string $visibility, bool $static, bool $byRef): self
```
Comment (dunglas) [inline src/Symfony/Component/PropertyInfo/PropertyReadInfo.php]: I would just provide a standard constructor instead of these factory method, for consistency with the rest of the code base.

File: src/Symfony/Component/PropertyInfo/PropertyReadInfoExtractorInterface.php [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+namespace Symfony\Component\PropertyInfo;
+
+/**
+ * Extract read information for the property of a class.
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ *
+ * @internal
```
Comment (dunglas) [inline src/Symfony/Component/PropertyInfo/PropertyReadInfoExtractorInterface.php]: Is it really necessary to make this class internal? It can be useful outside of Symfony itself.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/PropertyAccess/PropertyAccessor.php [hunk 28]
```diff
@@ -376,17 +387,22 @@ private function readProperty(array $zval, string $property, bool $ignoreInvalid
...
+                $result[self::VALUE] = $object->{$access->getName()};
 
-            if ($access[self::ACCESS_REF] && isset($zval[self::REF])) {
-                $result[self::REF] = &$object->{$access[self::ACCESS_NAME]};
+                if (isset($zval[self::REF]) && $access->canBeReference()) {
+                    $result[self::REF] = &$object->{$access->getName()};
+                }
             }
-        } elseif (!$access[self::ACCESS_HAS_PROPERTY] && property_exists($object, $property)) {
+        } elseif ($object instanceof \stdClass && property_exists($object, $property)) {
```
[the remark this answers is not in the document]
Comment (Korbeil) [reply src/Symfony/Component/PropertyAccess/PropertyAccessor.php]: This `elseif` is only here for `\stdClass` cases, so we check if given object is an `\stdClass` before going in or not.

**Comment by Korbeil** [comment]
@lyrixx @dunglas ping, fixed all comments.

About errors, I add to update `PropertyWriteInfo` DTO to store errors and use theses in `PropertyAccessor` class :wink:

**Review by dunglas — APPROVED** [review]

**Review by lyrixx — APPROVED** [review]

**Comment by bastnic** [comment]
This is huge, thanks @Korbeil!

**Review by maxhelias — APPROVED** [review]

**Comment by lyrixx** [comment]
Thank you @joelwurtz.
